File: apps/api-node/src/ingestion.ts
import { createHash } from "node:crypto";
import { Decimal } from "decimal.js";

export type Source = "LEDGER" | "COUNTERPARTY";
type CanonicalData = Record<string, string>;
type Normalizer = (raw: Record<string, string>) => CanonicalData;

export type Adapter = {
  id: string;
  source: Source;
  description: string;
  headers: readonly string[];
  normalize: Normalizer;
};

export class FileValidationError extends Error {
  constructor(public errors: any[]) {
    super("file validation failed");
  }
}

function decimal(value: string, field: string) {
  try {
    const result = new Decimal(value.trim());
    if (result.isNegative() || !result.isFinite()) throw new Error();
    return result.toString();
  } catch {
    throw new Error(`${field} must be a non-negative finite decimal`);
  }
}

function parseTime(value: string, assumeUtc = false) {
  const text = value.trim();
  const normalized =
    assumeUtc && !/[zZ]|[+-]\d\d:\d\d$/.test(text)
      ? text.replace(" ", "T") + "Z"
      : text;
  const date = new Date(normalized);
  if (Number.isNaN(date.getTime()))
    throw new Error("must be an ISO-8601 date-time");
  if (!assumeUtc && !/[zZ]|[+-]\d\d:\d\d$/.test(text))
    throw new Error("must include a timezone");
  return date.toISOString();
}

function mappedNormalizer(
  mapping: Record<string, string>,
  assumeUtc = false,
): Normalizer {
  return (raw) => {
    const value = (field: string) => raw[mapping[field]].trim();
    return {
      external_id: value("external_id"),
      executed_at: parseTime(value("executed_at"), assumeUtc),
      instrument: value("instrument").toUpperCase(),
      side: value("side").toUpperCase(),
      quantity: decimal(value("quantity"), mapping.quantity),
      price: decimal(value("price"), mapping.price),
      gross_amount: decimal(value("gross_amount"), mapping.gross_amount),
      state: value("state").toUpperCase(),
    };
  };
}

const ledgerHeaders = [
  "trade_id",
  "traded_at",
  "instrument",
  "side",
  "quantity",
  "price",
  "gross_amount",
  "state",
] as const;
const counterpartyHeaders = [
  "reference",
  "executed_at",
  "symbol",
  "direction",
  "qty",
  "unit_price",
  "total",
  "status",
] as const;
const canonicalHeaders = [
  "transaction_id",
  "executed_at",
  "instrument",
  "side",
  "quantity",
  "unit_price",
  "gross_amount",
  "state",
] as const;
const canonicalMapping = {
  external_id: "transaction_id",
  executed_at: "executed_at",
  instrument: "instrument",
  side: "side",
  quantity: "quantity",
  price: "unit_price",
  gross_amount: "gross_amount",
  state: "state",
};

export const adapters: readonly Adapter[] = [
  {
    id: "ledger-v1",
    source: "LEDGER",
    description: "Original ledger export",
    headers: ledgerHeaders,
    normalize: mappedNormalizer({
      external_id: "trade_id",
      executed_at: "traded_at",
      instrument: "instrument",
      side: "side",
      quantity: "quantity",
      price: "price",
      gross_amount: "gross_amount",
      state: "state",
    }),
  },
  {
    id: "counterparty-v1",
    source: "COUNTERPARTY",
    description: "Original counterparty statement",
    headers: counterpartyHeaders,
    // This format needs code because its compact B/S values are not simple renames.
    normalize: (raw) => ({
      external_id: raw.reference.trim(),
      executed_at: parseTime(raw.executed_at, true),
      instrument: raw.symbol.trim().toUpperCase(),
      side:
        ({ B: "BUY", S: "SELL", BUY: "BUY", SELL: "SELL" } as any)[
          raw.direction.trim().toUpperCase()
        ] ?? "",
      quantity: decimal(raw.qty, "qty"),
      price: decimal(raw.unit_price, "unit_price"),
      gross_amount: decimal(raw.total, "total"),
      state: raw.status.trim().toUpperCase(),
    }),
  },
  {
    id: "ledger-canonical-v1",
    source: "LEDGER",
    description: "Canonical column export",
    headers: canonicalHeaders,
    normalize: mappedNormalizer(canonicalMapping),
  },
  {
    id: "counterparty-canonical-v1",
    source: "COUNTERPARTY",
    description: "Canonical column export",
    headers: canonicalHeaders,
    normalize: mappedNormalizer(canonicalMapping),
  },
];

export const adaptersFor = (source: Source) =>
  adapters.filter((adapter) => adapter.source === source);

// This small RFC-4180 row parser is sufficient for generated take-home fixtures.
function csvLine(line: string) {
  const output: string[] = [];
  let current = "";
  let quoted = false;
  for (let index = 0; index < line.length; index++) {
    const character = line[index];
    if (character === '"' && line[index + 1] === '"' && quoted) {
      current += '"';
      index++;
    } else if (character === '"') quoted = !quoted;
    else if (character === "," && !quoted) {
      output.push(current);
      current = "";
    } else current += character;
  }
  output.push(current);
  return output;
}

export function resolveAdapter(
  content: Buffer,
  source: Source,
  requested?: string,
) {
  const candidates = adaptersFor(source);
  if (requested) {
    const selected = candidates.find((adapter) => adapter.id === requested);
    if (!selected)
      throw new FileValidationError([
        {
          row: 1,
          column: "adapter_id",
          value: requested,
          reason: "adapter is not registered for this source",
        },
      ]);
    return selected;
  }
  const actual = csvLine(
    content
      .toString("utf8")
      .replace(/^\uFEFF/, "")
      .split(/\r?\n/)[0] ?? "",
  );
  const matches = candidates.filter(
    (adapter) =>
      JSON.stringify(adapter.headers) ===
      JSON.stringify(actual.slice(0, adapter.headers.length)),
  );
  if (matches.length !== 1) {
    const expected = candidates
      .map((item) => item.headers.join(","))
      .join(" or ");
    throw new FileValidationError([
      {
        row: 1,
        column: "header",
        value: actual.join(","),
        reason: matches.length
          ? "format is ambiguous"
          : `unsupported file format; expected ${expected}`,
      },
    ]);
  }
  return matches[0];
}

export function parseCsv(content: Buffer, source: Source, adapterId?: string) {
  const text = content.toString("utf8").replace(/^\uFEFF/, "");
  const adapter = resolveAdapter(content, source, adapterId);
  const lines = text.split(/\r?\n/).filter(Boolean);
  const actual = csvLine(lines[0] ?? "");
  if (
    JSON.stringify(actual.slice(0, adapter.headers.length)) !==
    JSON.stringify(adapter.headers)
  )
    throw new FileValidationError([
      {
        row: 1,
        column: "header",
        value: actual.join(","),
        reason: `expected ${adapter.headers.join(",")}`,
      },
    ]);

  const rows: any[] = [];
  const errors: any[] = [];
  const seen = new Map<string, string>();
  for (let index = 1; index < lines.length; index++) {
    const values = csvLine(lines[index]);
    const raw = Object.fromEntries(
      actual.map((header, column) => [header, values[column] ?? ""]),
    );
    try {
      const data: CanonicalData & { source: Source } = {
        source,
        ...adapter.normalize(raw),
      };
      if (!data.external_id) throw new Error("transaction ID is required");
      if (!data.instrument) throw new Error("instrument is required");
      if (!["BUY", "SELL"].includes(data.side))
        throw new Error("side must be BUY/SELL or B/S");
      if (!["SETTLED", "CANCELLED"].includes(data.state))
        throw new Error("state must be SETTLED or CANCELLED");
      const canonical = JSON.stringify(
        Object.fromEntries(Object.entries(data).sort()),
      );
      if (
        seen.has(data.external_id) &&
        seen.get(data.external_id) !== canonical
      )
        throw new Error("conflicting duplicate transaction ID");
      if (!seen.has(data.external_id))
        rows.push({
          data,
          raw,
          fingerprint: createHash("sha256").update(canonical).digest("hex"),
        });
      seen.set(data.external_id, canonical);
    } catch (error: any) {
      errors.push({
        row: index + 1,
        column: "row",
        value: JSON.stringify(raw),
        reason: error.message,
      });
    }
  }
  if (errors.length) throw new FileValidationError(errors);
  if (!rows.length)
    throw new FileValidationError([
      {
        row: 0,
        column: "file",
        value: "",
        reason: "file contains no data rows",
      },
    ]);
  return rows;
}

File: apps/api-python/app/ingestion.py
import csv
import hashlib
import io
import json
from dataclasses import dataclass
from typing import Callable

from .domain import decimal_value, parse_time


@dataclass(frozen=True)
class Adapter:
    id: str
    source: str
    description: str
    headers: tuple[str, ...]
    normalize: Callable[[dict[str, str]], dict]


@dataclass
class FileValidationError(Exception):
    errors: list[dict]


def _mapped_normalizer(mapping: dict[str, str], *, assume_utc: bool = False):
    """Build a declarative adapter for formats that only rename columns."""

    def normalize(raw: dict[str, str]) -> dict:
        def value(field: str) -> str:
            return raw[mapping[field]].strip()

        return {
            "external_id": value("external_id"),
            "executed_at": parse_time(value("executed_at"), assume_utc).isoformat(),
            "instrument": value("instrument").upper(),
            "side": value("side").upper(),
            "quantity": str(decimal_value(value("quantity"), mapping["quantity"])),
            "price": str(decimal_value(value("price"), mapping["price"])),
            "gross_amount": str(
                decimal_value(value("gross_amount"), mapping["gross_amount"])
            ),
            "state": value("state").upper(),
        }

    return normalize


def _counterparty_v1(raw: dict[str, str]) -> dict:
    """The counterparty format needs a code adapter for B/S side aliases."""
    side = {"B": "BUY", "S": "SELL", "BUY": "BUY", "SELL": "SELL"}.get(
        raw["direction"].strip().upper(), ""
    )
    return {
        "external_id": raw["reference"].strip(),
        "executed_at": parse_time(raw["executed_at"], True).isoformat(),
        "instrument": raw["symbol"].strip().upper(),
        "side": side,
        "quantity": str(decimal_value(raw["qty"], "qty")),
        "price": str(decimal_value(raw["unit_price"], "unit_price")),
        "gross_amount": str(decimal_value(raw["total"], "total")),
        "state": raw["status"].strip().upper(),
    }


CANONICAL_HEADERS = (
    "transaction_id",
    "executed_at",
    "instrument",
    "side",
    "quantity",
    "unit_price",
    "gross_amount",
    "state",
)
CANONICAL_MAPPING = {
    "external_id": "transaction_id",
    "executed_at": "executed_at",
    "instrument": "instrument",
    "side": "side",
    "quantity": "quantity",
    "price": "unit_price",
    "gross_amount": "gross_amount",
    "state": "state",
}

ADAPTERS = (
    Adapter(
        "ledger-v1",
        "LEDGER",
        "Original ledger export",
        (
            "trade_id",
            "traded_at",
            "instrument",
            "side",
            "quantity",
            "price",
            "gross_amount",
            "state",
        ),
        _mapped_normalizer(
            {
                "external_id": "trade_id",
                "executed_at": "traded_at",
                "instrument": "instrument",
                "side": "side",
                "quantity": "quantity",
                "price": "price",
                "gross_amount": "gross_amount",
                "state": "state",
            }
        ),
    ),
    Adapter(
        "counterparty-v1",
        "COUNTERPARTY",
        "Original counterparty statement",
        (
            "reference",
            "executed_at",
            "symbol",
            "direction",
            "qty",
            "unit_price",
            "total",
            "status",
        ),
        _counterparty_v1,
    ),
    Adapter(
        "ledger-canonical-v1",
        "LEDGER",
        "Canonical column export",
        CANONICAL_HEADERS,
        _mapped_normalizer(CANONICAL_MAPPING),
    ),
    Adapter(
        "counterparty-canonical-v1",
        "COUNTERPARTY",
        "Canonical column export",
        CANONICAL_HEADERS,
        _mapped_normalizer(CANONICAL_MAPPING),
    ),
)


def adapters_for(source: str) -> list[Adapter]:
    return [adapter for adapter in ADAPTERS if adapter.source == source]


def _decode(content: bytes) -> str:
    try:
        return content.decode("utf-8-sig")
    except UnicodeDecodeError:
        raise FileValidationError(
            [{"row": 0, "column": "file", "value": "", "reason": "must be UTF-8"}]
        )


def resolve_adapter(content: bytes, source: str) -> Adapter:
    candidates = adapters_for(source)
    header = tuple(next(csv.reader(io.StringIO(_decode(content))), []))
    matches = [
        item for item in candidates if header[: len(item.headers)] == item.headers
    ]
    if len(matches) != 1:
        expected = " or ".join(",".join(item.headers) for item in candidates)
        reason = (
            "format is ambiguous"
            if matches
            else f"unsupported file format; expected {expected}"
        )
        raise FileValidationError(
            [
                {
                    "row": 1,
                    "column": "header",
                    "value": ",".join(header),
                    "reason": reason,
                }
            ]
        )
    return matches[0]


def parse_csv(content: bytes, source: str):
    """Validate a complete file and return canonical rows without side effects."""
    text = _decode(content)
    adapter = resolve_adapter(content, source)
    reader = csv.DictReader(io.StringIO(text))
    actual_headers = tuple(reader.fieldnames or ())
    if actual_headers[: len(adapter.headers)] != adapter.headers:
        raise FileValidationError(
            [
                {
                    "row": 1,
                    "column": "header",
                    "value": ",".join(reader.fieldnames or []),
                    "reason": f"expected {','.join(adapter.headers)}",
                }
            ]
        )

    rows, errors, seen = [], [], {}
    for number, parsed in enumerate(reader, start=2):
        # Preserve every source column while making ragged rows ordinary errors.
        raw = {column: (parsed.get(column) or "") for column in actual_headers}
        try:
            data = {"source": source, **adapter.normalize(raw)}
            if not data["external_id"]:
                raise ValueError("transaction ID is required")
            if not data["instrument"]:
                raise ValueError("instrument is required")
            if data["side"] not in ("BUY", "SELL"):
                raise ValueError("side must be BUY/SELL or B/S")
            if data["state"] not in ("SETTLED", "CANCELLED"):
                raise ValueError("state must be SETTLED or CANCELLED")
            canonical = json.dumps(data, sort_keys=True, separators=(",", ":"))
            external_id = data["external_id"]
            if external_id in seen and seen[external_id] != canonical:
                raise ValueError("conflicting duplicate transaction ID")
            if external_id not in seen:
                fingerprint = hashlib.sha256(canonical.encode()).hexdigest()
                rows.append((data, raw, fingerprint))
            seen[external_id] = canonical
        except (ValueError, KeyError) as exc:
            errors.append(
                {
                    "row": number,
                    "column": "row",
                    "value": json.dumps(raw),
                    "reason": str(exc),
                }
            )
    if errors:
        raise FileValidationError(errors)
    if not rows:
        raise FileValidationError(
            [
                {
                    "row": 0,
                    "column": "file",
                    "value": "",
                    "reason": "file contains no data rows",
                }
            ]
        )
    return rows
